shared/api/v3/course.ts

```diff
@@ -0,0 +1,160 @@
+
+type integer = number;
+
+export interface Course {
+    courseCode: string;
+    courseName: string;
+    courseSortKey: string[];
+    courseMutualExclusionKey: string[];
+    courseDescription: string;
+    courseInstructors: string[];
+    courseTerm: string;
+    courseSchedule: Schedule[];
+    courseCredits: number;
+    courseSeatsTotal: integer;
+    courseSeatsFilled: integer;
+    courseWaitlistLength: null;
+    courseEnrollmentStatus: string;
+    permCount: integer;
+}
+
+export interface Schedule {
+    scheduleDays: string;
+    scheduleStartTime: string;
+    scheduleEndTime: string;
+    scheduleStartDate: string;
+    scheduleEndDate: string;
+    scheduleTermCount: integer;
+    scheduleTerms: integer[];
+    scheduleLocation: string;
+}
+
+export function courseFromV4Section(s: APIv4.Section): Course {
+    const sectionCodeString = APIv4.stringifySectionCode(s.identifier);
+    const courseCodeString = APIv4.stringifyCourseCode(s.course.code);
+
+    const startDate = stringifyCourseDate(s.startDate);
+    const endDate = stringifyCourseDate(s.endDate);
+
+    const { termCount, terms } = getSectionTermFractions(s.identifier);
+
+    return {
+        courseCode: sectionCodeString,
+        courseName: s.course.title,
+        courseSortKey: [sectionCodeString],
+        courseMutualExclusionKey: [courseCodeString],
+        courseDescription: s.course.description,
+        courseInstructors: s.instructors.map((inst) => inst.name),
+        courseTerm: stringifySectionTermCode(s.identifier),
+        courseSchedule: s.schedules.flatMap((schedule) => {
+            const days = stringifyWeekdayEnumArray(schedule.days);
+            const startTime = stringifyTime(schedule.startTime);
+            const endTime = stringifyTime(schedule.endTime);
+            return schedule.locations.map((location) => ({
+                scheduleDays: days,
+                scheduleStartTime: startTime,
+                scheduleEndTime: endTime,
+                scheduleStartDate: startDate,
+                scheduleEndDate: endDate,
+                scheduleTermCount: termCount,
+                scheduleTerms: terms,
+                scheduleLocation: location,
+            }));
+        }),
+        courseCredits: s.credits,
+        courseSeatsTotal: s.seatsTotal,
+        courseSeatsFilled: s.seatsFilled,
+        courseWaitlistLength: null,
+        courseEnrollmentStatus: stringifySectionStatusEnum(s.status),
+        permCount: s.permCount,
+    };
+}
+
+export function courseListFromV4SectionList(sections: APIv4.Section[]): {
+    terms: [];
+    courses: Course[];
+} {
+    return {
+        terms: [],
+        courses: sections.map(courseFromV4Section),
+    };
+}
+
+function stringifyWeekdayEnumArray(arr: APIv4.WeekdayEnum[]): string {
+    return arr.join("");
+}
+
+function stringifyTime(secs: integer): string {
+    if (secs < 0 || secs >= 24 * 60 * 60) {
+        throw Error(`Time out of range: ${secs}`);
```

Comment:
Since your function can throw errors, be sure to add error handling to the function you will write to convert a list of v4 sections to v3 sections. If there is any error processing a section, log it and continue with the rest, and discard the invalid sections. We don't want crash the entire server just because we can't parse a singe course section.